PyQt6 Upgrade

i1sm3ky opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
This PR aims to upgrade Vorta from using PyQt5 to PyQt6.

### Related Issue
Issue: #765 PyQt6 Upgrade by @m3nu.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This change upgrades Vorta's GUI library (PyQt5) to its latest version at the time (PyQt6) which natively supports ARM architecture. Due to an update on PyQt5, it is no longer supported on ARM architecture which makes us unable to install and use Vorta on ARM based operating systems. By upgrading Vorta from PyQt5 to PyQt6, it'll add support for ARM architecture allowing us to use Vorta on ARM based systems too.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been tested using the pre-written unit tests.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


*I provide my contribution under the terms of the [license](./../../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

<!--
This template is sourced from the awesome https://github.com/TalAter/open-source-templates
-->

m3nu commented:
Oh, I think you accidentally pushed all your icon work too. The PyQt6 commit should be based on the current master branch, not your other work. Now it's all mixed up. Best to force-push this again.

m3nu commented:
You can just cherry-pick your one PyQt commit, as tested here: https://github.com/borgbase/vorta/pull/1686

Commands:

- `git checkout master`  (also be sure you pull the latest changes
- `git cherry-pick c97f3e970e606d379a7828f2bed8e606bb9dac01` (hash of the one commit you want)
- (then fix a minor conflict)
- done

i1sm3ky commented:
> Oh, I think you accidentally pushed all your icon work too. The PyQt6 commit should be based on the current master branch, not your other work. Now it's all mixed up. Best to force-push this again.

I initially made the changes on my fork which contained the changes for the icons too and the icons also took some changes to work with PyQt6 so I thought those could also be added here. I'll make those changes on the icons PR and keep this to PyQt6 upgrade only.

m3nu commented:
Great, thanks! You should also change `pyqt5` in `setup.cfg`. After that it works without issue.

I'm surprised it doesn't need more changes than this. Was expecting to update the .UI files too.

If this is really the case, we should quickly start a new 0.9 branch with this.

i1sm3ky commented:
> Great, thanks! You should also change `pyqt5` in `setup.cfg`. After that it works without issue.

Ya, I forgot to change that.

> I'm surprised it doesn't need more changes than this. Was expecting to update the .UI files too.

The only major changes it took were the full enums and the imports, for which I wrote a script that took care of those. Other than that it required some minor changes here and there including change in the import of QAction and QShortcut, removal of High DPI scaling, some flag changes, data type changes in SetChecked, change in some deprecated methods and some more.

> If this is really the case, we should quickly start a new 0.9 branch with this.

Yes, after fixing those failing tests which I mentioned I think we might be good with the upgrade.

m3nu commented:
Yeah, I see the changes. It's a bit more than search & replace, but not a crazy change. And no change to the UI XML files.

For your commits, you added one new commit that undoes previous commits. Would be better to just have the actual changes. You see all the commits of this PR here: https://github.com/borgbase/vorta/pull/1685/commits

If you like I can also force-push just your one commit to this repo and we take it from there. But I'm sure you can also figure it out.

m3nu commented:
A system package may need adding in .github/workflows/test.yml

https://packages.ubuntu.com/bionic/libegl1

i1sm3ky commented:
> If you like I can also force-push just your one commit to this repo and we take it from there. But I'm sure you can also figure it out.

You can force-push my commit to this repo if you want. My bad I've committed my icons work on my master branch so when I do checkout master it still keeps those changes.

m3nu commented:
Done. As you can see we only have 2 commits now and the history is much clearer than first adding and then removing stuff.

But don't worry, this Git stuff can be confusing in the beginning.

i1sm3ky commented:
> Done. As you can see we only have 2 commits now and the history is much clearer than first adding and then removing stuff.
> 
> But don't worry, this Git stuff can be confusing in the beginning.

Yeah, Thanks for the help! But it still contains some of my code from the icon issue, is that fine?

m3nu commented:
With this Ubuntu package added, test properly run on all platforms, but have a few minor errors. Can look into those tomorrow.

i1sm3ky commented:
> With this Ubuntu package added, test properly run on all platforms, but have a few minor errors. Can look into those tomorrow.

Yeah Sure!

m3nu commented:
Packaging with PyInstaller for macOS also works well. The resulting app is actually 30% or so smaller than Qt5. 🤷‍♂️

i1sm3ky commented:
> Packaging with PyInstaller for macOS also works well. The resulting app is actually 30% or so smaller than Qt5. 🤷‍♂️

Oh great! I think the upgrade will be complete sooner then expected.

real-yfprojects commented on src/vorta/views/extract_dialog.py:
```diff
@@ -616,7 +616,7 @@ def set_checkstate_recursively(self, index: QModelIndex, value: Qt.CheckState):
         item = index.internalPointer()
         for i in range(number_children):
-            child = index.child(i, 0)
+            child = index.sibling(i, 0)
```
`sibling` is not the same as `child`. This is the reason for the one of the test failures.

i1sm3ky replied:
`child` is deprecated now. The docs recommend to use `QAbstractItemModel.index(row, column[, parent=QModelIndex()])` but when I use it, it gives
```
TypeError: index(self, row: int, column: int, parent: QModelIndex = QModelIndex()): first argument of unbound method must have type 'QAbstractItemModel'
```

real-yfprojects replied:
Use
```suggestion
            child = self.index(i, 0, index)
```

i1sm3ky replied:
Yeah it fixed it. Thanks!

i1sm3ky replied:
Should we add the color changing icons after PyQt6 upgrade? So that we won't need to update it to fix for PyQt6 afterwards.

diivi replied:
Some testing might be required for the PyQt6 upgrades, so maybe merging that and updating this branch afterwards would be better.

i1sm3ky replied:
It won't be that long before we could push this update I think. But either way I've a PR opened for the current fix and I might also change it to make it compatible with the PyQt6 as it is without requiring much change in future.

diivi replied:
There are many other PRs open, too. Updating this PR would be better than updating all of them to be compatible with pyqt6. Anyway, this is a bigger fix than that one and can wait on another branch as Manu said.

i1sm3ky replied:
Yeah, I made the required chances to the Icons PR. That can now be added anytime before or after the PyQt6 upgrade. It works on both, just requires to change the imports which this PR would be doing to all the imports anyway.

m3nu replied:
In most cases, there shouldn't be many conflicts with existing PRs after merging the qt6 update. Since most changes are just to imports. And I prefer to be done with this update and have future work build on top of it instead of maintaining 2 versions for too long. And we merged many finished PRs for the next release.

So I'll do the 0.8.11 release and add a new 0.8 branch that will only get bugfixes.

The master branch will get this qt6 PR and the version 0.9.x in the future.

Will look into the failing tests now.

real-yfprojects commented on tests/test_extract.py:
```diff
@@ -88,20 +88,20 @@ def test_selection():
 
     # Test select
-    ic = model.index(1, 0, QModelIndex())
+    ic = model.sibling(1, 0, QModelIndex())
```
`index` is not the same as `sibling`.

i1sm3ky replied:
Reverted those to the original `index`. My bad changed those by mistake, Thanks for pointing it out.

m3nu commented:
Fixed the tests. Nothing major to change. There were some subtle changes to dBus notifications. Would be good if someone can test those on Linux. 🙏

i1sm3ky commented:
Should I mark this PR as ready for review?

m3nu commented:
> Should I mark this PR as ready for review?

Yes, please!

i1sm3ky commented:
What should I do/work on next?

m3nu commented:
> What should I do/work on next?

If you're on Linux, you can test this PR with Qt6 a bit more. E.g. ensure DBus notifications are working.

i1sm3ky commented:
> If you're on Linux, you can test this PR with Qt6 a bit more. E.g. ensure DBus notifications are working.

Yeah sure! How should I go on testing DBus notifications?

i1sm3ky commented:
> If you're on Linux, you can test this PR with Qt6 a bit more. E.g. ensure DBus notifications are working.

Yeah sure! How should I go on testing the DBus notifications?

m3nu commented:
Run some backups and see if e.g. you see notifications for errors or success (given the setting is enabled in the *Misc* tab). Of if there are any display issues or crashes when using a dialog, etc.

i1sm3ky commented:
> Run some backups and see if e.g. you see notifications for errors or success (given the setting is enabled in the _Misc_ tab). Of if there are any display issues or crashes when using a dialog, etc.

Yup gotcha! I can test those.

diivi commented:
Good News:
Now qt respects dark mode on my device out of the box!
Bad News:
```bash
QDBusMarshaller: type `QMetaType::Type' (65589) is not registered with D-BUS. Use qDBusRegisterMetaType to register it
2023-04-09 19:21:38,337 - vorta.notifications - WARNING - Failed to send notification!
2023-04-09 19:21:38,337 - vorta.notifications - WARNING - Type of message, “(sisssa{sv}i)”, does not match expected type “(susssasa{sv}i)”
```
The notifications don't seem to work.

diivi commented:
Vorta from the command line is also not working (I hope I am not doing something wrong):
![image](https://user-images.githubusercontent.com/41837037/230777246-1d594dc3-36c5-4c6f-8eb7-e63407e16a24.png)

m3nu commented:
Looks like some more adjustments are needed. Can you add a comment in the part of the code that fails to keep track?

diivi commented:
https://github.com/borgbase/vorta/pull/1685/files#diff-2fc212f20eb85d55fed0805e9deaabc1a8a75ddbec54fda3c4d246cb0a1e572aR55 for the command line errors.

And all notifications.

diivi commented:
Hovering over the `i` icons in the Misc tab here:
![image](https://user-images.githubusercontent.com/41837037/230777663-3d717a49-272c-4785-b206-d27a6da1bf62.png)
Raises this error:
![image](https://user-images.githubusercontent.com/41837037/230777672-34da6762-5393-4cd1-839a-d100607f8bab.png)

diivi commented:
Trying to delete an archive in the Archives tab:
![image](https://user-images.githubusercontent.com/41837037/230777771-46b51eb5-bf72-4a94-9496-914fc63bc9d3.png)

i1sm3ky commented:
> Hovering over the `i` icons in the Misc tab here:

globalPos() is deprecated, use globalPosition().

m3nu commented:
With the new feature to save the sort order of the source tab: Qt now saves a string instead of an int:

```
CRITICAL - Uncaught exception, file a report at https://github.com/borgbase/vorta/issues/new/choose
Traceback (most recent call last):
  File "__main__.py", line 78, in <module>
  File "__main__.py", line 71, in main
  File "application.py", line 72, in __init__
  File "views/main_window.py", line 56, in __init__
  File "views/source_tab.py", line 97, in __init__
  File "views/source_tab.py", line 250, in populate_from_profile
ValueError: invalid literal for int() with base 10: 'SortOrder.AscendingOrder'
```

So we need a short check to translate this to an int.

![Screenshot 2023-04-09 at 15 20 47](https://user-images.githubusercontent.com/3916435/230778173-780acb6a-70ab-452d-b63d-2dfd39ecc526.jpg)

m3nu commented:
Thanks! If you see an issue that's quick to fix, just do it.

i1sm3ky commented:
> Trying to delete an archive in the Archives tab:

Fixed it. It seems we need to check and convert more enums to their fully qualified names, and change some of pending and hidden exec_()s to exec().

i1sm3ky commented:
> > Trying to delete an archive in the Archives tab:
> 
> Fixed it. It seems we need to check and convert more enums to their fully qualified names, and change some of pending and hidden exec_()s to exec().

Most of them would be in `QMessageBox`s I guess.

m3nu commented:
Great! If you fix something, also push the commit, so I can easily add changes on top.

i1sm3ky commented:
> Great! If you fix something, also push the commit, so I can easily add changes on top.

Yeah sure!

m3nu commented:
Fixed the issue with saving table sorting.

m3nu commented:
Fixed the headless issue by removing the encoding option (UTF-8) is already the default as per https://doc.qt.io/qt-6/qtextstream.html#setEncoding

m3nu commented:
> globalPos() is deprecated, use globalPosition(). (I'll update those)

Still getting an error after changing this:

```
 CRITICAL - Uncaught exception, file a report at https://github.com/borgbase/vorta/issues/new/choose
Traceback (most recent call last):
  File "/Users/manu/Workspace/vorta/src/vorta/views/partials/tooltip_button.py", line 101, in mouseMoveEvent
    QToolTip.showText(event.globalPosition(), self.toolTip(), self)
TypeError: showText(pos: QPoint, text: str, widget: typing.Optional[QWidget] = None, rect: QRect = {}, msecShowTime: int = -1): argument 1 has unexpected type 'QPointF'
```

m3nu commented:
It needs `toPoint()`. Fixed.

https://doc.qt.io/qt-6/qpointf.html#details

i1sm3ky commented:
> It needs `toPoint()`. Fixed.
> 
> https://doc.qt.io/qt-6/qpointf.html#details

i1sm3ky commented:
Sorry for that! It happened by accident.

i1sm3ky commented:
> Still getting an error after changing this:

It was working on mine.

> It needs `toPoint()`. Fixed.
> 
> https://doc.qt.io/qt-6/qpointf.html#details

Oh thanks! Great to know that.

i1sm3ky commented:
I'll do further tests on linux after I resolve some issues I'm having with cloning it on Linux.

diivi commented:
I pulled the latest branch from Kunal's repository.
@m3nu I think your change didn't work? Without changing anything and running Vorta - 

![image](https://user-images.githubusercontent.com/41837037/230782402-1c1419c8-3fa9-43d3-afda-b84a2060cd40.png)

m3nu commented:
> @m3nu I think your change didn't work? Without changing anything and running Vorta -

You get this error because you already have a wrong value in your settings database.

So you need to open `settings.db` in the `Vorta` folder (location depends on OS) and make sure there are only `int`s in the `settingsmodel` table.

![Screenshot 2023-04-09 at 15 20 47](https://user-images.githubusercontent.com/3916435/230782785-c5d147f5-9805-455b-81c9-c6b6da1af506.jpg)

m3nu commented:
We need to check all Enum-to-int comparison. This e.g. broke our Extract feature.

More here: https://stackoverflow.com/a/75052368/3983708

Example:
- not working any more: `item.data.checkstate == Qt.CheckState.Checked`
- working: `item.data.checkstate == Qt.CheckState.Checked.value`

i1sm3ky commented:
> We need to check all Enum-to-int comparison. This e.g. broke our Extract feature.

Yes, we'll need to check other enums as well, Most of them are fixed but some of them are still not converted to their full qualified names.

m3nu commented:
Seeing some issues with Qt in testing today. Maybe after their 6.5 update. Will test a bit locally.

Also worked on DBus notifications. Our call may need some simplifications. Found a code snippet here: https://bugreports.qt.io/browse/PYSIDE-1905

```
============================= test session starts ==============================
platform linux -- Python 3.7.16, pytest-7.3.0, pluggy-1.0.0 -- /opt/hostedtoolcache/Python/3.7.16/x64/bin/python
cachedir: .pytest_cache
PyQt6 6.5.0 -- Qt runtime 6.5.0 -- Qt compiled 6.5.0
rootdir: /home/runner/work/vorta/vorta
configfile: setup.cfg
testpaths: tests
plugins: mock-3.10.0, qt-4.2.0, cov-4.0.0
collecting ... collected 126 items

tests/test_archives.py::test_prune_intervals qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: linuxfb, xcb, minimal, offscreen, vnc, eglfs, minimalegl, wayland, vkkhrdisplay, wayland-egl.

Fatal Python error: Aborted
```

i1sm3ky commented:
What else is left to do maybe I could help in some of them? (I won't be able to do any linux based ones as I'm unable to install PyQt6 or any PyQt on ARM based linux on my m1, Once I get home in summer then or when my friend lets me use his device I could test some stuff on there otherwise I'm unable to test for linux as of now :/)

m3nu commented:
Didn't we do this upgrade to support Arm in the first place?

Some QMessageBoxes are blocking tests on macOS since the update. Hopefully fixed now.

m3nu commented:
> What else is left to do maybe I could help in some of them?

Testing and making sure enums are fixed, @i1sm3ky. 🙏

Else I managed to take care of a few annoying issues today. So getting there..

i1sm3ky commented:
> Didn't we do this upgrade to support Arm in the first place?

Yes we did, and it's working fine on MacOS. It has something to do with how linux handles its packages. I think as I'm using Ubuntu, It doesn't have complete support and functionality for its ARM version as of now.

i1sm3ky commented:
> Testing and making sure enums are fixed, @i1sm3ky. 🙏

Ok, I'll check all of them once. 
> Else I managed to take care of a few annoying issues today. So getting there..

Great!

i1sm3ky commented:
> We need to check all Enum-to-int comparison. This e.g. broke our Extract feature.
> 
> More here: https://stackoverflow.com/a/75052368/3983708
> 
> Example:
> 
> * not working any more: `item.data.checkstate == Qt.CheckState.Checked`
> * working: `item.data.checkstate == Qt.CheckState.Checked.value`

Found some more Enum-to-int comparisons and assignments in `extract_dialog.py`. In forms of `.data.checkstate == Qt.CheckState.Unchecked` and `.data.checkstate = value` where `value` is 
```
(parameter) value: Literal[CheckState.Unchecked, CheckState.PartiallyChecked, CheckState.Checked]
```

Adding the value attribute to those to make them int-to-int comparison and assignment made the `test_extract` fail due to int-to-enum comparison. When I converted the comparisons in test to int-to-int, it passed.
So should I keep the changes or leave it as it was? (I haven't committed the changes yet, LMK if you want to see what I changed, I'll commit those)

Other than that I checked all the files for correct enums along with all the other known and found changes in PyQt6 and no other issue was found.

m3nu commented:
Good stuff. Please push your changes, they sound reasonable.

diivi commented:
> Are we happy to merge this, @real-yfprojects @diivi ? It's fine to iron out smaller bugs later. Just want to avoid blocking other large issues, like your formatting PR.

My general workflow seems to work except the issues I mentioned in the thread (which you and @i1sm3ky have fixed).

So yeah, if I missed testing a small detail, it could be fixed later. 
I like the automatic dark theme that I got with this update, excited to do development with it.

i1sm3ky commented:
After this gets merged, I'll test the icons once and get it merged.
After all this, what should I work on next?

i1sm3ky commented:
Once this gets merged, I'll test the icons once and get it merged.
After all this, what should I work on next?

diivi commented:
> what should I work on next?

You can pick any open issue and start working on it; the mentors will gladly help! 
Also, for GSoC, I'd recommend checking out other projects from the Borg Collective too (Borgmatic & Borg) if you like.

Otherwise, if you want a specific issue, maybe the mentors can guide you.

i1sm3ky commented:
> You can pick any open issue and start working on it; the mentors will gladly help! Browsing the tracker will also help you gain experience with the project, understanding what issues we face currently etc. Also, for GSoC, I'd recommend checking out other projects from the Borg Collective too (Borgmatic & Borg) if you like.
> 
> Otherwise, if you want a specific issue, maybe the mentors can guide you.

Thanks for the reply. I know those, I was just asking so that if there's any critical issue right now or something they think is important right now which need to be done soon, I could work on that!

m3nu commented:
Got some errors in the *Extract* dialog. So changed it a bit to use `Qt.CheckState` everywhere to avoid mixing it up with ints.

i1sm3ky commented:
> Got some errors in the _Extract_ dialog. So changed it a bit to use `Qt.CheckState` everywhere to avoid mixing it up with ints.

Oh great! May I ask what was the error about? I'm curious what went wrong.

m3nu commented:
Confusion of types that would throw exceptions when selecting files. You can see my commit for details. Hope this is the last enum error to surface.

real-yfprojects reviewed (commented):
This has to be updated too, I think: https://github.com/borgbase/vorta/blob/6edd4091aaf98eb8ee8260bd102bf423847f5f3f/setup.cfg#L103

I couldn't test notifications because they no longer work on my testing machine for some reason.

m3nu commented on src/vorta/tray_menu.py:
```diff
@@ -73,6 +73,14 @@ def build_menu(self):
         exit_action = menu.addAction(self.tr('Quit'))
         exit_action.triggered.connect(self.app.quit)
 
+    def is_taskbar_dark(self):
+        app = QApplication.instance()
+        long_color = QScreen.grabWindow(app.primaryScreen()).toImage().pixel(0, 0)
+        int_color = int(long_color)
+        return 150 > QColor.lightness(
+            QColor.fromRgb(((int_color >> 16) & 0xFF), ((int_color >> 8) & 0xFF), (int_color & 0xFF))
+        )
```
True, please remove this, @i1sm3ky . Then re-add it in another PR.

m3nu replied:
Never mind, already done.

m3nu commented on src/vorta/views/extract_dialog.py:
```diff
@@ -532,7 +532,7 @@ def data(self, index: QModelIndex, role: int = Qt.ItemDataRole.DisplayRole):
             return item.data.checkstate
 
-    def setData(self, index: QModelIndex, value, role: int = Qt.ItemDataRole.DisplayRole) -> bool:
+    def setData(self, index: QModelIndex, value: Qt.CheckState, role: int = Qt.ItemDataRole.DisplayRole) -> bool:
```
Right. Anything that could be improved? Should we have a hard check to validate the type? Like for role.

real-yfprojects replied:
I don't think that's needed. We don't do type checks in any other python code. However the type hints should be updated. While value can remain untyped in the signature (it could be cast to `CheckState` later though), role might not be of type `int` any more. Someone has to check that and any other methods expecting a qt enum.

i1sm3ky replied:
> role might not be of type `int` any more. Someone has to check that and any other methods expecting a qt enum.

The Qt6 docs states `int role = Qt::DisplayRole)` it still is of type int in PyQt6.

m3nu replied:
The comparison works with ints, yes. I do think we mostly get `role` as type `Qt.ItemDataRole.DisplayRole` and then compare it to be the same. So shouldn't need any change.

```
In [7]: Qt.ItemDataRole.DisplayRole == 0
Out[7]: True
```

m3nu replied:
I do wonder if this would be a better type hint?

`def data(self, index: QModelIndex, role: Qt.ItemDataRole = Qt.ItemDataRole.DisplayRole):`

i1sm3ky replied:
We use roles mostly as `DisplayRole`, then I think it makes sense.

real-yfprojects replied:
```suggestion
    def setData(self, index: QModelIndex, value, role: Union[int, Qt.ItemDataRole] = Qt.ItemDataRole.DisplayRole) -> bool:
```

will be best since Qt passes `int` for `role` while devs (we) might want to pass `ItemDataRole`.

i1sm3ky replied:
> will be best since Qt passes `int` for `role` while devs (we) might want to pass `ItemDataRole`.

Ya you're right, that would be more helpful as a type hint.

m3nu replied:
Pushed

real-yfprojects replied:
The Or for type hints was introduced with py3.10 though.

m3nu replied:
Oops. Shows up in the tests.

m3nu commented on src/vorta/views/extract_dialog.py:
```diff
@@ -281,7 +281,7 @@ class FileData:
     source_path: Optional[str] = None  # only relevant for links
 
-    checkstate: int = 0  # whether to extract the file (0, 1 or 2)
+    checkstate: Qt.CheckState = Qt.CheckState.Unchecked  # whether to extract the file (0, 1 or 2)
```
You mean comparison with int? I hope we changed all of them to compare `CheckState` to `CheckState` by now.

real-yfprojects replied:
When I wrote this review there where multiple occurrences of `item.data.checkstate == state.value` or similar.

real-yfprojects replied:
Or maybe there was only one which I have promptly fixed now.

m3nu commented:
> This has to be updated too, I think:
> 
> https://github.com/borgbase/vorta/blob/6edd4091aaf98eb8ee8260bd102bf423847f5f3f/setup.cfg#L103
> 
> I couldn't test notifications because they no longer work on my testing machine for some reason.

Changed. Do we still use pylint anywhere?

m3nu commented:
> I couldn't test notifications because they no longer work on my testing machine for some reason.

Notifications should work. I fixed and tested them on Ubuntu 22.04.

m3nu commented:
I've pinned pyobjc, since today's update causes a segfault in our autostart code. This will be a separate PR.

real-yfprojects commented:
> Changed. Do we still use pylint anywhere?

I use it locally. But it isn't used in the repository.

m3nu commented:
Merged this and will soon do a v0.9.0 beta release so macOS users can easily test the compiled version.

Thanks to @i1sm3ky for getting the ball rolling on this and @real-yfprojects for reviewing and adding valuable fixes and input!

m3nu commented:
You can look for stale PRs that may only need a small push or change to be useable. E.g. #1639 should be nearly ready, but was abandoned by the person or something.

Careful, since some PRs may also include features that are either too specific or add too much maintenance effort in the long term. So we don't want to merge those.

i1sm3ky commented:
> You can look for stale PRs that may only need a small push or change to be useable. E.g. #1639 should be nearly ready, but was abandoned by the person or something.

Ok thanks! I had seen this issue before, I'll work on it and get it fixed.

> Careful, since some PRs may also include features that are either too specific or add too much maintenance effort in the long term. So we don't want to merge those.

 Ok gotcha.

m3nu commented:
These docs may need updating too. I'm on an Intel mac and can't test Arm stuff.

https://vorta.borgbase.com/contributing/#local-development-setup

Would be good to build an Arm binary or universal binary somehow. GH Actions is also Intel.

i1sm3ky commented:
> These docs may need updating too. I'm on an Intel mac and can't test Arm stuff.
> 
> https://vorta.borgbase.com/contributing/#local-development-setup
> 
> Would be good to build an Arm binary or universal binary somehow. GH Actions is also Intel.

I'll look more into it and get back to you.

i1sm3ky commented:
> Would be good to build an Arm binary or universal binary somehow. 

`pyinstaller --clean --noconfirm --target-arch universal2 package/vorta.spec` It should build a universal binary which could run on both -x86_64 as well as arm64 architecture.

real-yfprojects commented:
I think you need universal2 python binaries for that to work.

i1sm3ky commented:
> Not impossible to build universal2 Python bins:
> 
> [pyenv/pyenv#2262 (comment)](https://github.com/pyenv/pyenv/issues/2262#issuecomment-1498347584)
> 
> Maybe an afternoon to get this working on Github Actions with macOS.

The Apple dev docs suggest a similar approach using lipo to merge the individual binaries.

i1sm3ky commented:
> I think you need universal2 python binaries for that to work.

Thanks, I think you're right, I might have to look further into it.
